### PR TITLE
[WIP][RFC] Edit nn.Linear to take arbitrary in and out feature dimensions

### PR DESCRIPTION
**Pull request by mrshenli**
### Current Behavior

Currently, `nn.Linear` takes two integer parameters `in_features` and `out_features` and creates a weight parameter of size `[out_features, in_features]`, and bias parameter of size `[out_features]`. This API is a little cumbersome if users would like to run higher-dimensional features through a linear layer. Say I have 2D input features of size `[2, 3]`, and provide an input batch of size `[4, 5, 2, 3]`. To use the existing linear layer, I will have to first reshape the the input into size `[4, 5, 6]`, and pass `6` as `in_features` arg for `nn.Linear`. 

### Proposed Behavior

It would be nice if `nn.Linear` can handle arbitrary feature dimensions. For example, if I have input feature size `[2, 3]`, and output feature size `4`, I should be able to do the following:

```python
m = Linear(in_features=[2, 3], out_features=4)
input = torch.rand(4, 5, 2, 3)
output = m(input) # get output of size [4, 5, 4]
```

Let `in_features=[in_1, in_2, ..., in_x]`, `out_features=[out_1, out_2, ..., out_y]`, and `input.size() == [d1, d2, ..., dz]`. The output dimensions should be `[d_1, ..., d_{z-x}, out_1, ..., out_y]`. 

It is allowed to set `in_features` and/or `out_features` to `[]` or `()`, and its behavior will stay consistent with `torch.matmul`.

#### Option 1 [implemented in this PR]

Weights and bias dimensions respect `in_features` and `out_features` provided by users, where `weight.size() == [out_1, out_2, ..., out_y, in_1, in_2, ..., in_x]`  and `bias.size() == out_features`. In every `forward()` iteration, `weight` is reshaped to a 2D tensor of size `[prod(out_features), prod(in_features)]`, and `bias` of size `[prod(in_features)]`. Then, inputs are reshaped to `[d_1, ..., d_{z-x}, prod(in_features)]`. `F.linear` function creates an intermediate output of size `[d_1, ..., d_{z-x}, prod(out_features)]`, and are reshaped to `[d_1, ..., d_z, out_1, ..., out_y]`. 

The reason for not keeping weight and bias views as member var is to avoid showing them in `Linear.parameters()` 

Pro: Dimensions shown in `Linear.parameters()` will match users' expectation.
Con: Have to create multiple views in every `forward`.

#### Option 2

In constructor, directly create weight of size `[prod(out_features), prod(in_features)]` and bias of size `[prod(out_features)]`. And only reshape input and output in every `forward()` iteration. 

Pro: Avoid reshaping `weight` and `bias` in `forward()`
Con: Dimensions shown in `Linear.parameters()` won't match `in_features` and `out_features`

#### Option 3
Implement a new `linear` native function which address the dimensions internal, and modify both Python and C++ frontend to call it.

Will edit docs and add comments when we have consensus.

@gchanan @vishwakftw @yf225 

### Benchmark [Python API only]

before

```python
In [2]: l = torch.nn.Linear(10, 10)

In [3]: input = torch.rand(2, 10)

In [4]: %timeit l(input)
26.2 µs ± 5.17 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [5]: %timeit l(input)
25.1 µs ± 1.57 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [6]: %timeit l(input)
25.6 µs ± 1.93 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

after

```python
In [2]: l = torch.nn.Linear(10, 10)

In [4]: input = torch.rand(2, 10)

In [5]: %timeit l(input)
24.4 µs ± 383 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [6]: %timeit l(input)
26.2 µs ± 4.25 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [7]: %timeit l(input)
24.8 µs ± 2.37 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

### CODE REVIEW

**Comment by vishwakftw** [comment]
I think `view`ing might slightly expensive. On a related note, are there any scenarios where the intermediate tensors cannot be viewed in the current implementation?

**Comment by mrshenli** [comment]
> On a related note, are there any scenarios where the intermediate tensors cannot be viewed in the current implementation?

@vishwakftw Which tensors?

**Comment by vishwakftw** [comment]
> @vishwakftw Which tensors?

Few instances where viewing which can possibly throw errors are these
- https://github.com/pytorch/pytorch/pull/20258/files#diff-531f4c06f42260d699f43dabdf741b6dR132
- https://github.com/pytorch/pytorch/pull/20258/files#diff-b3ce6f6da2faf48ee488d334b9833f34R62

Is the output of `F.linear` always guaranteed to be contiguous?

(Sorry about the links; I don't know why permalink isn't available)

File: torch/csrc/api/src/nn/modules/linear.cpp [hunk 1]
```diff
@@ -36,7 +52,29 @@ void LinearImpl::pretty_print(std::ostream& stream) const {
 
 Tensor LinearImpl::forward(const Tensor& input) {
   AT_ASSERT(!options.with_bias_ || bias.defined());
-  return torch::linear(input, weight, bias);
+  auto input_view = input.view_as(input);
```
Comment (vishwakftw) [inline torch/csrc/api/src/nn/modules/linear.cpp]: Why is this necessary? Sorry if I missed something obvious.

Comment (mrshenli) [reply torch/csrc/api/src/nn/modules/linear.cpp]: This is not required. If the cost of creating a view is a concern, i will edit the code to avoid this. 

I was assuming that view adds negligible overhead, and hence creating a view upfront for less LoC.

**Comment by ezyang** [comment]
Is there an issue associated with this PR?

File: torch/nn/modules/linear.py [hunk 2]
```diff
@@ -73,23 +76,70 @@ def __init__(self, in_features, out_features, bias=True):
...
+        self.weight = Parameter(torch.empty(sizes))
         if bias:
-            self.bias = Parameter(torch.Tensor(out_features))
+            self.bias = Parameter(torch.empty(self.out_features))
         else:
             self.register_parameter('bias', None)
         self.reset_parameters()
 
+    def _sizes_prod(self, features):
+        assert isinstance(features, (int, Iterable)), (
```
Comment (ezyang) [inline torch/nn/modules/linear.py]: Don't use an assert here. Assert means developer (us) error, but in this case your checking if the user passed in something bad.

File: torch/nn/modules/linear.py [hunk 3]
```diff
@@ -73,23 +76,70 @@ def __init__(self, in_features, out_features, bias=True):
...
+
+    def _reshape_params(self):
+        if self.in_prod != self.in_features or self.out_prod != self.out_features:
+            weight_view = self.weight.view(self.out_prod, self.in_prod)
+            bias_view = self.bias.view(self.out_prod)
+            return weight_view, bias_view
+        else:
+            return self.weight, self.bias
 
     @weak_script_method
```
[the remark this answers is not in the document]
Comment (ezyang) [reply torch/nn/modules/linear.py]: It almost certainly does, since you didn't declare `_reshape_params` as a script method (at the very least)

File: torch/nn/modules/linear.py [hunk 4]
```diff
@@ -73,23 +76,70 @@ def __init__(self, in_features, out_features, bias=True):
...
     def forward(self, input):
-        return F.linear(input, self.weight, self.bias)
```
Comment (ezyang) [inline torch/nn/modules/linear.py]: Is there a reason the logic was done here, as opposed to adjusting the internal implementation of linear to be generalized in this way? Because you did it this way, you had to also duplicate the logic in the C++ module implementation. If you had put it in linear, the C++ module would have gotten the logic as well.

**Comment by ezyang** [comment]
In your PR body, you say:

> The last len(in) dimensions of input data are feature dimensions, and all preceding dimensions are batch dimensions.

What is `in`?

**Comment by ezyang** [comment]
For my information, what is an example use case for this feature? It's not obvious to me under what circumstances you'd be willing to remove all of the structure of the input dimensions and just matrix multiply it as if it were flat.

**Comment by ezyang** [comment]
I tried to review the logic change, but I couldn't figure out what the intended semantics were supposed to be. Furthermore, the documentation in the codebase was not updated, so I could not read it to see what the semantics were. And I didn't understand the PR body description (see above).

**Comment by gchanan** [comment]
> The last len(in) dimensions of input data are feature dimensions, and all preceding dimensions are batch dimensions. For input with dimensions [d_1, ..., d_n, in_1, ..., in_x], the output dimensions will be [d_1, ..., d_n, out_1, ..., out_y]. The behavior stays consistent with torch.matmul. If out_features=[], output dimensions will be [d_1, ..., d_n] instead of [d_1, ..., d_n, 1].

This is pretty difficult to understand.  This doesn't define whether you are referring to current behavior or new behavior.  Terms are also not defined or are used inconsistently (`in` is used before being defined, `out_features` presumably refers to the output analog of `in` but isn't called `out` (to disambiguate it from an `out=` parameter?).  How about an example?

**Comment by gchanan** [comment]
> I would vote for this one as benchmarks show no regression.

You haven't said what the alternative is yet.

File: torch/nn/modules/linear.py (hunk 4 above)
Comment (mrshenli) [reply torch/nn/modules/linear.py]: I might miss something, it seems to me the existing code has two implementations already: [python](https://github.com/pytorch/pytorch/blob/master/torch/nn/functional.py#L1390-L1412) and [c++](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/Linear.cpp#L14-L28). I can submit another PR to merge these two first, if that is the preferred solution. 

Regarding why not doing it in the `linear` function, I think we can but will need to add some more args to it to specify how many in and out feature dimensions.  Thoughts?

**Comment by mrshenli** [comment]
@ezyang 

> What is in?

I mean the `in_features` arg in `nn.Linear`.

> For my information, what is an example use case for this feature? It's not obvious to me under what circumstances you'd be willing to remove all of the structure of the input dimensions and just matrix multiply it as if it were flat.

Let me edit the PR summary to clarify this.

File: test/test_nn.py [hunk 5]
```diff
@@ -7192,6 +7192,73 @@ def m(t):
...
+        input_sizes.append(prod(in_list))
+
+        output = torch.matmul(input.view(input_sizes), weight.t()) + bias
+
+        output_sizes = list(output.size())[:-1]
+        output_sizes.extend(out_list)
+        output = output.view(output_sizes)
+
+        self.assertEqual(output, m(input))
+        self.assertEqual(torch.Size(expected_output_sizes), m(input).size())
```
Comment (gchanan) [inline test/test_nn.py]: can this ever fail if the check above passes?  I guess if you pass in the wrong expected_output_sizes, but why do you need to pass that in anyway?

File: torch/nn/modules/linear.py (hunk 4 above)
Comment (gchanan) [reply torch/nn/modules/linear.py]: I agree with @ezyang's point here.

**Review by gchanan — CHANGES_REQUESTED** [review]

File: torch/nn/modules/linear.py (hunk 4 above)
Comment (gchanan) [reply torch/nn/modules/linear.py]: @bddppq can you weigh in here?  I thought we were going to unify these, or do I have that wrong?

Comment (bddppq) [reply torch/nn/modules/linear.py]: Thanks for notification. Yes we are trying to directly call the c++ aten::native::linear in the python side. For everyone who is not familiar with the background:
Currently we have two separated but same implementations of linear at [python side F.linear](https://github.com/pytorch/pytorch/blob/7b733e4fc12145583ef96a699cfef5e6b30eee54/torch/nn/functional.py#L1390-L1412) and [c++ side aten::native::linear](https://github.com/pytorch/pytorch/blob/7b733e4fc12145583ef96a699cfef5e6b30eee54/aten/src/ATen/native/Linear.cpp#L14-L28). python module nn.Linear module calls python `F.linear`, so technically c++ linear is not called by any place. Current setup causes some troubles for other backend integration (e.g. mkldnn and tvm, see explanation here https://github.com/pytorch/pytorch/pull/19210#issuecomment-486330064) and we are working on changing python `F.linear` to directly call c++ `aten::native::linear`. @wanchaol has couple onging PRs to make this change.
I would suggest wait for @wanchaol's PRs to land, and then change c++ linear to support the new semantics. If you do not want to wait then you would need to implement the new semantics in both python F.linear and c++ aten linear. In both cases, the nn.Linear's forward function should be kept unchanged.

File: test/test_nn.py (hunk 5 above)
Comment (mrshenli) [reply test/test_nn.py]: The test basically implements `linear` again by directly calling `matmul`. If the test's implementation of `linear` is incorrect, it is possible that `output` matches `m(input)` but both are wrong.

Let me add some tests cases that sets linear weight and bias, and then manually craft expected output values.

**Comment by mrshenli** [comment]
PR Summary updated
